New post: Debugging memory leaks in WebAssembly (#3664)
GoogleChrome/web.dev: src/site/content/en/blog/webassembly-memory-debugging/index.md

  result.delete();
  // module.doLeakCheck();

  return imgData;
  return new ImageData(result, result.width, result.height);
}
```

This also means we no longer need a custom `free_result` binding on the C++ side:

```cpp//0-3,5-9,13
void free_result(uint8_t* result) {
  free(result);
}

EMSCRIPTEN_BINDINGS(my_module) {
  class_<RawImage>("RawImage")
      .property("buffer", &RawImage::buffer)
      .property("width", &RawImage::width)
      .property("height", &RawImage::height);

  function("quantize", &quantize);
  function("zx_quantize", &zx_quantize);
  function("version", &version);
  function("free_result", &free_result, allow_raw_pointers());
}
```

All in all, our wrapper code became both cleaner and safer at the same time.

After this I went through some further minor improvements to the code of the ImageQuant wrapper and
replicated similar memory management fixes for other codecs. If you're interested in more details,
you can see the resulting PR here: [Memory fixes for C++
codecs](https://github.com/GoogleChromeLabs/squoosh/pull/780).

## Takeaways

What lessons can we learn and share from this refactoring that could be applied to other codebases?

> - Don't use memory views backed by WebAssembly—no matter which language it's built from - beyond a

Author: ```suggestion
- Don't use memory views backed by WebAssembly—no matter which language it's built from—beyond a
```

---

> title: Debugging memory leaks in WebAssembly

Reviewer: ```suggestion
title: Debugging memory leaks in WebAssembly using Emscripten
```

Last minute drive-by suggestion: I feel like we should mention “Emscripten” somewhere in the title as the article doesn’t really help you debug memory leaks when you are _not_ using Emscripten. Sound reasonable?

Author: Heh, nitially it said "...in WebAssembly wrappers" but I removed "wrappers" as the title felt a bit long. I guess we can put this in, but it makes it longer again.

Do you think this is an important piece of information?

Author: If it is, I wonder if at that point we should remove "WebAssembly" from the title altogether and make it "Debugging memory leaks with Emscripten".

Reviewer: It’s a bigger thing I have been thinking about. TL;DR is that I think tooling is at the core of the WebAssembly story and I think it’s important that we mention in the title that this is Emscripten-centric. If you feel like my suggestion is too long, I am also totally okay with dropping “WebAssembly” from the title.

Author: > TL;DR is that I think tooling is at the core of the WebAssembly story and I think it’s important that we mention in the title that this is Emscripten-centric.

Totally agree with that reasoning. I guess my concern is that people who don't use Emscripten, will skip the article, even though the API design principles apply equally to any language, only the sanitizer part is Emscripten-specific.

Reviewer: I personally don’t think that a non-Emscripten user will get a lot from this article. I totally see that the principles are transferrable, but without tooling support, that’s gonna be hard to do. So I’d vote for making this one Emscripten specific and link to (or write) articles for other tooling setups.

Author: Heh, perhaps adding sanitizers to the post wasn't a great idea after all. Initially this was meant to be a  language-agnostic post about safer API design for wrappers, but adding the complementary sanitizers part made it indeed very Emscripten-specific since only Emscripten implements the ASAN runtime for Wasm for now (this might change in future).

Fine, I'll rename the title, and will at least try to "fix" this by encouraging people to read it nevertheless in a tweet-announcement.

subhead: While JavaScript is fairly forgiving in cleaning up after itself, static languages are definitely not…
description: Learn how to use WebAssembly to bring libraries, written in other languages, to the Web in a safe and idiomatic manner.
date: 2020-08-06

> hero: hero.jpg
> authors:

Reviewer 2: ```suggestion
hero: hero.jpg
alt: TODO
authors:
```
Please add an `alt` description for the hero image using this property

  - rreverser
tags:
  - blog # blog is a required tag for the article to show up in the blog.
  - webassembly

>   - security

Reviewer: Use-After-Free, the anti-pattern that this article helps you spot and fix, is one of the main sources of security attacks in Chrome.

That being said, I could see how we want to reserve the “security” tag for security-specific APIs. Additionally, Wasm is sandboxed, so it would be about application-level security at the worst.

I’ll leave it up to you to make a call.

>   - security

Author: Yeah I don't have a strong opinion either. I've assumed that "security" tag applied to best security practices for Web apps as well and not just Chrome-wide security posts. If I understood correctly, then this class of bugs we're describing falls into that category, but if not, feel free to remove.

>   - security

Reviewer 2: The connection between use-after-free and security is a good enough rationale IMO for leaving `security` on the post, thanks for the extra context

  attr = liq_attr_create();
  image = liq_image_create_rgba(attr, image_buffer, image_width, image_height, 0);
  liq_set_max_colors(attr, num_colors);
  liq_image_quantize(image, attr, &res);
  liq_set_dithering_level(res, dithering);
  uint8_t* image8bit = (uint8_t*)malloc(size);
  result = (uint8_t*)malloc(size * 4);

  // …

  free(image8bit);
  liq_result_destroy(res);
  liq_image_destroy(image);
  liq_attr_destroy(attr);

  return {
    val(typed_memory_view(image_width * image_height * 4, result)),
    image_width,
    image_height
  };
}

void free_result() {
  free(result);
}
```

JavaScript (well, TypeScript):

```ts
export async function process(data: ImageData, opts: QuantizeOptions) {
  if (!emscriptenModule) {
    emscriptenModule = initEmscriptenModule(imagequant, wasmUrl);
  }
  const module = await emscriptenModule;

  const result = module.quantize(/* … */);

  module.free_result();

  return new ImageData(
    new Uint8ClampedArray(result.view),
    result.width,
    result.height
  );
}
```

Do you spot an issue? Hint: it's
[use-after-free](https://owasp.org/www-community/vulnerabilities/Using_freed_memory), but in
JavaScript!

> In Emscripten, `typed_memory_view` returns a JavaScript `Uint8Array` backed by the WebAssembly

Author: Good point!

memory buffer, with `byteOffset` and `byteLength` set to the given pointer and length. The main
point is that this is a TypedArray *view* into a WebAssembly memory buffer, rather than a
JavaScript-owned copy of the data.

When we call `free_result` from JavaScript, it, in turn, calls a standard C function `free` to mark
this memory as available for any future allocations, which means the data that our `Uint8Array` view
points to, can be overwritten with arbitrary data by any future call into Wasm.

Or, some implementation of `free` might even decide to zero-fill the freed memory immediately. The
`free` that Emscripten uses doesn't do that, but we are relying on an implementation detail here
that cannot be guaranteed.

Or, even if the memory behind the pointer gets preserved, new allocation might need to grow the
WebAssembly memory. When `WebAssembly.Memory` is grown either via JavaScript API, or corresponding
`memory.grow` instruction, it invalidates the existing `ArrayBuffer` and, transitively, any views
backed by it.

> Let me use DevTools or Node.js console to demonstrate this behaviour:

Reviewer 2: ```suggestion
Let me use the DevTools (or Node.js) console to demonstrate this behavior:
```

The sentence looks a bit grammatically off without `the`

Style guide tells us to use American spelling: https://developers.google.com/style/spelling

keep the application running regardless.

Let's expose that second helper via Embind so that we can call it from JavaScript at any time:

```cpp/0,12/
#include <sanitizer/lsan_interface.h>

// …

void free_result() {
  free(result);
}

EMSCRIPTEN_BINDINGS(my_module) {
  function("zx_quantize", &zx_quantize);
  function("version", &version);
  function("free_result", &free_result);
  function("doLeakCheck", &__lsan_do_recoverable_leak_check);
}
```

And invoke it from the JavaScript side once we're done with the image. Doing this from the
JavaScript side, rather than the C++ one, helps to ensure that all the scopes have been
exited and all the temporary C++ objects were freed by the time we run those checks:

```typescript/7/
  // …

  const result = opts.zx
    ? module.zx_quantize(data.data, data.width, data.height, opts.dither)
    : module.quantize(data.data, data.width, data.height, opts.maxNumColors, opts.dither);

  module.free_result();
  module.doLeakCheck();

  return new ImageData(
    new Uint8ClampedArray(result.view),
    result.width,
    result.height
  );
}
```

This gives us a report like the following in the console:

> ![Screenshot of a message "Direct leak of 12 bytes" coming from an anonymous wasm-function[…]](leak1-nodbg.png)

Reviewer 2: Please use this formatting: https://web.dev/handbook/markup-media/#partial-screenshots

Uh-oh, there are some small leaks, but the stacktrace is not very helpful as all the function names
are mangled. Let's recompile with a basic debugging info to preserve them:

```shell/12/
emcc \
  --bind \
  ${OPTIMIZE} \
  --closure 1 \
  -s ALLOW_MEMORY_GROWTH=1 \
  -s MODULARIZE=1 \
  -s 'EXPORT_NAME="imagequant"' \
  -I node_modules/libimagequant \
  -o ./imagequant.js \
  --std=c++11 \
  imagequant.cpp \
  -fsanitize=address \
  -g2 \
  node_modules/libimagequant/libimagequant.a
```

This looks much better:

> ![Screenshot of a message "Direct leak of 12 bytes" coming from a `GenericBindingType<RawImage>::toWireType` function](leak1-dbg.png)

Reviewer 2: Please use this formatting: https://web.dev/handbook/markup-media/#partial-screenshots

to manually free any memory and call destructors from the JavaScript side once you're done with the
object. For Embind specifically, the [official
docs](https://emscripten.org/docs/porting/connecting_cpp_and_javascript/embind.html#memory-management)
suggest to call a `.delete()` method on exposed C++ classes:

> JavaScript code must explicitly delete any C++ object handles it has received, or the Emscripten
> heap will grow indefinitely.
>
> ```js
> var x = new Module.MyClass;
> x.method();
> x.delete();
> ```

Indeed, when we do that in JavaScript for our class:

```ts/7/
  // …

  const result = opts.zx
    ? module.zx_quantize(data.data, data.width, data.height, opts.dither)
    : module.quantize(data.data, data.width, data.height, opts.maxNumColors, opts.dither);

  module.free_result();
  result.delete();
  module.doLeakCheck();

  return new ImageData(
    new Uint8ClampedArray(result.view),
    result.width,
    result.height
  );
}
```

The leak goes away as expected.

### Discovering more issues with sanitizers

Building other Squoosh codecs with sanitizers reveals both similar as well as some new issues. For
example, I've got this error in MozJPEG bindings:

> ![Screenshot of a message "memory access out of bounds" coming from a `jpeg_start_compress` function](out-of-bounds.png)

Reviewer 2: Please use this formatting: https://web.dev/handbook/markup-media/#partial-screenshots

result into a potentially random memory address that happened to be stored in those variables at the
time of the call!

```cpp/0-1
uint8_t* output;
unsigned long size;
// …
jpeg_mem_dest(&cinfo, &output, &size);
```

Zero-initialising both variables before the invocation solves this issue, and now the code reaches a
memory leak check instead. Luckily, the check passes successfully, indicating that we don't have any
leaks in this codec.

## Issues with shared state

…Or do we?

We know that our codec bindings store some of the state as well as results in global static
variables, and MozJPEG has some particularly complicated structures.

```cpp
uint8_t* last_result;
struct jpeg_compress_struct cinfo;

val encode(std::string image_in, int image_width, int image_height, MozJpegOptions opts) {
  // …
}
```

What if some of those get lazily initialized on the first run, and then improperly reused on future
runs? Then a single call with a sanitizer would not report them as problematic.

Let's try and process the image a couple of times by randomly clicking at different quality levels
in the UI. Indeed, now we get the following report:

> ![Screenshot of a message "Direct leak of 262144 bytes" coming from a `jpeg_finish_compress` function](leak2.png)

Reviewer 2: Please use this formatting: https://web.dev/handbook/markup-media/#partial-screenshots

Here, it's not a leak, but us writing to a memory outside of the allocated boundaries 😱

Digging into the code of MozJPEG, we find that the problem here is that `jpeg_mem_dest`—the

> function that we use to allocate a memory destination for JPE  [reuses existing values of

Author: ```suggestion
function that we use to allocate a memory destination for JPEG—[reuses existing values of
```
